cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/helpers/OrderingHelper.kt

```diff
@@ -0,0 +1,165 @@
+    // in which order to analyze the functions), we remove them from the dependencies (as a key
+    // but also in the value of other functions' dependencies because we will definitely analyze
+    // it before).
+    val dependencies = prepareCallGraph(functions)
+
+    val orderedList = mutableListOf<Node>()
+
+    addFunctionsWithoutDependency(orderedList, dependencies)
+
+    // All remaining nodes still have some unfulfilled dependencies. We make some heuristics
+    // based on how many dependencies we cannot fulfill. We therefore first collect all
+    // transitive dependencies.
+    val changed = dependencies.keys.toMutableList()
+
+    while (changed.isNotEmpty()) {
+        val k = changed.removeFirst()
+        val additionalValues =
+            dependencies[k]
+                ?.flatMap { dependencies.computeIfAbsent(it, ::identitySetOf) }
+                ?.toIdentitySet() ?: identitySetOf()
+        if (dependencies.computeIfAbsent(k, ::identitySetOf).addAll(additionalValues)) {
+            changed.addAll(dependencies.filterValues { k in it }.keys.filter { it !in changed })
+        }
+    }
+    dependencies.forEach { (k, v) -> v.remove(k) }
+
+    log.info("Ordering all functions according to their dependencies")
+
+    while (dependencies.isNotEmpty()) {
+        // We always try to find functions without any (unsatisfied) dependencies. That's
+        // obviously the best scenario because we can be sure that all prerequisites have
+        // been fulfilled.
+        var nextFunctions = dependencies.filterValues { it.isEmpty() }.keys
+        if (nextFunctions.isEmpty()) {
+            // Each function has at least one dependency :( Then, we pick the function with the
+            // smallest number of missing dependencies.
+            // TODO: A more sophisticated approach could improve the results here. E.g. least
+            //   dependencies but used in most other functions.
+            val mappedEntries = dependencies.entries.map { Pair(it, it.value.size) }
+            val minimum = mappedEntries.minOf { it.second }
+            nextFunctions =
+                identitySetOf(
+                    mappedEntries
+                        .filter { it.second == minimum }
+                        .minBy { it.first.key.name }
+                        .first
+                        .key
+                )
+        }
+
+        nextFunctions.forEach { nextFunction ->
+            // It's no longer needed in the map
+            dependencies.remove(nextFunction)
+            // It's no longer an unsatisfied dependency.
+            dependencies.forEach { (_, v) -> v.remove(nextFunction) }
+        }
+        orderedList.addAll(nextFunctions.sortedBy { it.name })
+    }
+
+    // We add all things which are not a function declaration to the end because they won't be
+    // called at a specific point in time (we hope)
+    orderedList.addAll(noFunction)
```

Comment:
This is rather tricky. Could be at the beginning or here... Not sure what's best